File: 02-Prompt-Engineering/prompt-engineering-typescript/src/examples/1-tipos-de-prompts/3.1-CoT-self-consistency.ts
import { ChatOpenAI } from "@langchain/openai";
import {
  loadEnvironment,
  getApiKey,
  getEndpoint,
  printLlmResult,
} from "../../utils/helpers";
import { get } from "http";

loadEnvironment();

async function main(): Promise<void> {
  const message1 = `
    Question: In an API endpoint that returns a list of users and their posts, the developer wrote:

  users := db.FindAllUsers()
  for _, u := range users {
    u.Posts = db.FindPostsByUserID(u.ID)
  }

  How many database queries will this code execute if there are N users?

  Generate 3 different reasoning paths step by step.
  At the end, summarize the answers and choose the most consistent one, ignoring outliers.
  If there are 3 different answers, ONLY reply: "I can't find a consistent answer". 
  `;

  const model = new ChatOpenAI({
    model: "gpt-4o",
    configuration: {
      baseURL: getEndpoint("GITHUB_MODELS_ENDPOINT"),
      apiKey: getApiKey("GITHUB_MODELS_TOKEN"),
    },
  });

  const response1 = await model.invoke(message1);

  printLlmResult(message1, response1);
}

main();

File: 02-Prompt-Engineering/1-tipos-de-prompt/prompt-engineering-python/examples/1-tipos-de-prompts/3.1-CoT-Self-consistency.py
from langchain_openai import ChatOpenAI
from dotenv import load_dotenv
from pydantic import SecretStr
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))

from utils.prompt_helpers import print_llm_result

load_dotenv()

message1 = """
  Question: In an API endpoint that returns a list of users and their posts, the developer wrote:

  users := db.FindAllUsers()
  for _, u := range users {
    u.Posts = db.FindPostsByUserID(u.ID)
  }

  How many database queries will this code execute if there are N users?

  Generate 3 different reasoning paths step by step.
  At the end, summarize the answers and choose the most consistent one, ignoring outliers.
  If there are 3 different answers, ONLY reply: "I can't find a consistent answer".
"""

model = ChatOpenAI(
    model="gpt-4o",
    base_url=os.environ["GITHUB_MODELS_ENDPOINT"],
    api_key=SecretStr(os.environ["GITHUB_MODELS_TOKEN"])
)

response1 = model.invoke(message1)
print_llm_result(message1, response1)